docs/manual/returns-receiver-checker.tex

```diff
@@ -0,0 +1,96 @@
+type, the Returns Receiver Checker injects an
+\refqualclass{common/returnsrcvr/qual}{This} annotation on the method's receiver
+parameter.  The presence of the polymorphic
+\refqualclass{common/returnsrcvr/qual}{This} qualifier on both the method's
+return and receiver type force their types to be equal.  Hence, Checker
+Framework's standard checking of type compatibility will discover cases where a
+method is not properly returning its receiver.  When a method with an
+\refqualclass{common/returnsrcvr/qual}{This} return type returns some value
+other than its receiver, the issue is reported as a
+\code{[return.type.incompatible]} warning.  Similarly, if a method with an
+\refqualclass{common/returnsrcvr/qual}{This} return type is overridden by a
+method lacking \refqualclass{common/returnsrcvr/qual}{This} on its return type,
+the checker reports an \code{[override.return.invalid]} warning.
+
+Additionally, an \code{[invalid.this.location]} warning will be emitted if an
+\refqualclass{common/returnsrcvr/qual}{This} annotation is written in any
+location other than on the return type of a method or in a downcast.
+
+
+\sectionAndLabel{Framework Support}{returns-receiver-checker-framework-support}
+
+\begin{figure}
+    \begin{Verbatim}
+    @AutoValue
+    abstract class Animal {
+        abstract String name();
+        abstract int numberOfLegs();
+        static Builder builder() {
+            return new AutoValue_Animal.Builder();
+        }
+        @AutoValue.Builder
+        abstract static class Builder {
+            abstract Builder setName(String value);
+            abstract Builder setNumberOfLegs(int value);
+            abstract Animal build();
+        }
+    }
+    \end{Verbatim}
+    \caption{A code example using AutoValue's Builder support.}
+    \label{fig-autovalue-builder}
+\end{figure}
+
+For programs using builder generation support from
+\href{https://github.com/google/auto/tree/master/value}{AutoValue} or
+\href{https://projectlombok.org/}{Project Lombok}, the Returns Receiver Checker
+will automatically inject \refqualclass{common/returnsrcvr/qual}{This}
+annotations on the return types of the fluent API methods of generated builders.
+Consider the code using AutoValue in Figure~\ref{fig-autovalue-builder}.  For
+this code, AutoValue automatically generates a class
+\<AutoValue\_Animal.Builder> that extends \<Animal.Builder> and implements its
+abstract methods.  The Returns Receiver Checker automatically adds a
+\refqualclass{common/returnsrcvr/qual}{This} annotation to the return type of
+\<Animal.Builder.setName> and \<Animal.Builder.setNumberOfLegs>, and also to the
+return types of those methods in the generated subclass.  The Returns Receiver
+Checker still runs to ensure the implementation of the generated subclass is
+consistent with the injected annotations.  Project Lombok Builders are supported
+in a similar fashion.
+
+To disable the checker's built-in framework support, use the
```

Comment:
To be honest, I don't recall why we added the ability to disable the framework support feature.  @kelloggm do you remember?  I seem to recall you thought it would be useful.  I could imagine disabling framework support might slightly speed up the checker, but this has never been measured.

Reply:
It is necessary to run the machine image sniping check (and only that check) on projects that use AutoValue/Lombok. In those cases, we weren't interested in misuse of the frameworks, just in violations of the properties we'd encoded in our stub files.

More generally, it's useful because object construction via builders is a special-case of what the object construction checker's type system (called methods) is useful for.

For the RR checker, it's not clear to me why having the option is necessary, since it doesn't ever issue errors because of its framework support.

Reply:
Good point.  I will go ahead and remove documentation of this option for the Returns Receiver Checker.  We can document it properly when we merge in the Object Construction Checker.